spring-integration-core/src/main/java/org/springframework/integration/store/SimpleMessageStore.java

```diff
@@ -255,7 +292,9 @@ public MessageGroup removeMessageFromGroup(Object groupId, Message<?> messageToR
 				SimpleMessageGroup group = this.groupIdToMessageGroup.get(groupId);
 				Assert.notNull(group, "MessageGroup for groupId '" + groupId + "' " +
 						"can not be located while attempting to remove Message from the MessageGroup");
-				group.remove(messageToRemove);
+				if (group.remove(messageToRemove)) {
+	                groupUpperBound.release();
```

Comment:
PS. I will try and load test this and the entire 4.2.0 code Monday and let you know how it goes